zcash_primitives/src/transaction/builder.rs

```diff
@@ -706,22 +755,172 @@ impl<P: consensus::Parameters, U: sapling::builder::ProverProgress> Builder<'_,
         output_prover: &OP,
         fee_rule: &FR,
     ) -> Result<BuildResult, Error<FR::Error>> {
-        let fee = self.get_fee(fee_rule).map_err(Error::Fee)?;
-        self.build_internal(
-            transparent_signing_set,
-            sapling_extsks,
-            orchard_saks,
-            rng,
-            spend_prover,
-            output_prover,
-            fee,
-        )
+        if self.build_config.is_coinbase() {
+            self.build_coinbase(rng, spend_prover, output_prover)
+        } else {
+            let fee = self.get_fee(fee_rule).map_err(Error::Fee)?;
+
+            self.build_standard(
+                transparent_signing_set,
+                sapling_extsks,
+                orchard_saks,
+                rng,
+                spend_prover,
+                output_prover,
+                fee,
+            )
+        }
+    }
+
+    #[cfg(feature = "circuits")]
+    fn build_coinbase<R: RngCore + CryptoRng, SP: SpendProver, OP: OutputProver, FE>(
+        self,
+        mut rng: R,
+        spend_prover: &SP,
+        output_prover: &OP,
+    ) -> Result<BuildResult, Error<FE>> {
+        use super::Authorized;
+
+        let consensus_branch_id = BranchId::for_height(&self.params, self.target_height);
+
+        let (miner_data, seq) = match self.build_config {
+            BuildConfig::Coinbase {
+                miner_data,
+                sequence,
+            } => (miner_data, sequence),
+            _ => panic!(
+                "the build_coinbase method should not be called without a coinbase build config"
+            ),
+        };
```

Comment:
Why not do the match in `build` and then call this method with `miner_data` (and `sequence` if it still exists)? Then `is_coinbase` would also not be needed.

Reply:
I originally wrote it as you described, but I couldn't find a way to satisfy the borrow checker since `miner_data` isn't `Copy`.

If I do the match in `build` and borrow `miner_data` from `self` there, then I can't call `build_coinbase` on `self` anymore because `build_coinbase` needs to consume `self`.

Reply:
That's also why I added `is_coinbase`.